aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -39,94 +50,418 @@ fn is_string_type(context: &SafeNativeContext, type_arg: &Type) -> SafeNativeRes
+enum SnapshotType {
+    U128,
+    U64,
+    String,
+}
+
+impl SnapshotType {
+    fn from_ty_arg(context: &SafeNativeContext, ty_arg: &Type) -> SafeNativeResult<Self> {
+        match ty_arg {
+            Type::U128 => Ok(Self::U128),
+            Type::U64 => Ok(Self::U64),
+            _ => {
+                // Check if the type is a string
+                if is_string_type(context, ty_arg)? {
+                    Ok(Self::String)
+                } else {
+                    // If not a string, return an error
+                    Err(SafeNativeError::Abort {
+                        abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+                    })
+                }
+            },
+        }
+    }
+
+    pub fn pop_snapshot_field_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        self.parse_snapshot_value_by_type(aggregator_snapshot_field_value(&safely_pop_arg!(
+            args, StructRef
+        ))?)
+    }
+
+    pub fn pop_snapshot_value_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        match self {
+            SnapshotType::U128 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u128))),
+            SnapshotType::U64 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u64) as u128)),
+            SnapshotType::String => {
+                let input = string_to_bytes(safely_pop_arg!(args, Struct))?;
+                Ok(SnapshotValue::String(input))
+            },
+        }
+    }
+
+    pub fn parse_snapshot_value_by_type(&self, value: Value) -> SafeNativeResult<SnapshotValue> {
+        // Simpler to wrap to be able to reuse safely_pop_arg functions
+        self.pop_snapshot_value_by_type(&mut VecDeque::from([value]))
+    }
+
+    pub fn create_snapshot_value_by_type(&self, value: SnapshotValue) -> SafeNativeResult<Value> {
+        match (self, value) {
+            (SnapshotType::U128, SnapshotValue::Integer(v)) => Ok(Value::u128(v)),
+            (SnapshotType::U64, SnapshotValue::Integer(v)) => Ok(Value::u64(u128_to_u64(v)?)),
+            (SnapshotType::String, value) => {
+                Ok(Value::struct_(Struct::pack(vec![Value::vector_u8(
+                    match value {
+                        SnapshotValue::String(v) => v,
+                        SnapshotValue::Integer(v) => to_utf8_bytes(v),
+                    },
+                )])))
+            },
+            // ty_arg cannot be Integer, if value is String
+            _ => Err(SafeNativeError::Abort {
+                abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+            }),
+        }
+    }
+}
+
+// ================= START TEMPORARY CODE =================
+// TODO: aggregator_v2 branch will introduce these in different places in code
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SnapshotValue {
```

Comment:
We are going to land aggregator_v2 branch on main in a week anyway. What are we benefitting by landing these stuff here in main now, and changing them again after landing aggregator_v2?